bundles/org.openhab.binding.pilight/src/main/java/org/openhab/binding/pilight/internal/PilightConnector.java

```diff
@@ -0,0 +1,281 @@
+                                callback.statusReceived(Collections.singletonList(status));
+                            }
+                        }
+                        line = in.readLine();
+                    }
+                }
+            } catch (IOException e) {
+                if (running) {
+                    logger.debug("Error in pilight listener thread", e);
+                }
+            }
+
+            logger.info("Disconnected from pilight server at {}:{}", config.getIpAddress(), config.getPort());
+
+            if (running) {
+                callback.updateThingStatus(ThingStatus.OFFLINE, ThingStatusDetail.NONE, null);
+                // empty line received (socket closed) or pilight stopped but binding
+                // is still running, try to reconnect
+                connect();
+            }
+        }
+    }
+
+    /**
+     * Tells the connector to refresh the configuration
+     */
+    public void refreshConfig() {
+        logger.trace("refreshConfig");
+        doSendAction(new Action(Action.ACTION_REQUEST_CONFIG));
+    }
+
+    /**
+     * Tells the connector to refresh the status of all devices
+     */
+    public void refreshStatus() {
+        logger.trace("refreshStatus");
+        doSendAction(new Action(Action.ACTION_REQUEST_VALUES));
+    }
+
+    /**
+     * Stops the listener
+     */
+    public void close() {
+        running = false;
+        disconnect();
+        interrupt();
+    }
+
+    private void disconnect() {
+        final @Nullable PrintStream printStream = this.printStream;
+        if (printStream != null) {
+            printStream.close();
+            this.printStream = null;
+        }
+
+        final @Nullable Socket socket = this.socket;
+        if (socket != null) {
+            try {
+                socket.close();
+            } catch (IOException e) {
+                logger.debug("Error while closing pilight socket", e);
+            }
+            this.socket = null;
+        }
+    }
+
+    private boolean isConnected() {
+        final @Nullable Socket socket = this.socket;
+        return socket != null && !socket.isClosed();
+    }
+
+    private void connect() {
+        disconnect();
+
+        int delay = 0;
+
+        while (!isConnected()) {
+            try {
+                logger.debug("pilight connecting to {}:{}", config.getIpAddress(), config.getPort());
+
+                Thread.sleep(delay);
+                Socket socket = new Socket(config.getIpAddress(), config.getPort());
+
+                Options options = new Options();
+                options.setConfig(true);
+
+                Identification identification = new Identification();
+                identification.setOptions(options);
+
+                // For some reason, directly using the outputMapper to write to the socket's OutputStream doesn't work.
+                PrintStream printStream = new PrintStream(socket.getOutputStream(), true);
```

Comment:
At this point it would be bad to close the print stream, as it is assigned to the class variable to be used in some other methods for the actual communication to the pilight daemon. The printstream is closed in the disconnect method. But I'll also add a "close" call to the Else branch.

Reply:
You're right, I overlooked the assignment to the field. You also need to handle the case if the IOException is raised.